src/Effect/Ref.purs

```diff
@@ -20,7 +20,40 @@ import Effect (Effect)
 -- | which holds a value of type `a`.
 foreign import data Ref :: Type -> Type
 
+type role Ref representational
+
 -- | Create a new mutable reference containing the specified value.
+-- |
+-- | ## Why This Type Signature Cannot Be `forall s. s -> Ref s`
```

Comment:
I'd prefer to avoid Markdown subheadings, because it requires us to make assumptions about the markup the compiler generates for API documentation. At the moment each declaration in the module is introduced with an `<h3>`, but this gives us an `<h2>`, so we have a larger heading under a smaller one, which is no good.

Regarding the actual content here, beginners will occasionally think that all Ref operations ought to be pure, so I think the fact that most Ref operations have to run in Effect is something that deserves discussing at the top of the module.

Reply:
That is: it will come across a bit strange to devote so much effort to explaining why `new` specifically is effectful when almost everything in this module is effectful as well and none of the other functions have had the same treatment.

Reply:
If you could split the docs changes into a separate PR we can get this one merged and discuss this further there?

Reply:
Yeah, will do. I should have done it that way in the first place.